presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java

```diff
@@ -0,0 +1,190 @@
+ * 1. The filter can be pushed down to the scan node.
+ * 2. The rows not matching the condition are not aggregated.
+ * 3. The IF() expression wrapper is removed.
+ */
+public class RewriteAggIfToAggFilter
+        implements Rule<AggregationNode>
+{
+    private static final Capture<ProjectNode> CHILD = newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .with(source().matching(project().capturedAs(CHILD)));
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isAggIfToFilterRewriteEnabled(session);
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        ProjectNode sourceProject = captures.get(CHILD);
+
+        Set<AggregationNode.Aggregation> aggregationsToRewrite = aggregationNode.getAggregations().values().stream()
+                .filter(aggregation -> ShouldRewriteAggregation(aggregation, sourceProject))
+                .collect(toSet());
+        if (aggregationsToRewrite.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Extract the VariableReferenceExpressions from these aggregations.
+        // Multiple aggregations may reference the same input. We use a set to dedup them so that we only do the rewrite once per input IF expression.
+        Set<VariableReferenceExpression> aggregationReferences =
+                aggregationsToRewrite.stream().map(a -> (VariableReferenceExpression) a.getArguments().get(0)).collect(toSet());
+
+        // Get the corresponding assignments in the input project.
+        Map<VariableReferenceExpression, RowExpression> sourceAssignments = sourceProject.getAssignments().entrySet().stream()
+                .filter(entry -> aggregationReferences.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        Assignments.Builder newAssignments = Assignments.builder();
+        // We don't remove the IF expression now in case the aggregation has other references to it. These will be cleaned up by other pruning rules.
+        newAssignments.putAll(sourceProject.getAssignments());
+
+        // Map from the aggregation reference to the IF condition reference.
+        Map<VariableReferenceExpression, VariableReferenceExpression> aggregationReferenceToConditionReference = new HashMap<>();
+        // Map from the aggregation reference to the IF result reference.
+        Map<VariableReferenceExpression, VariableReferenceExpression> aggregationReferenceToIfResultReference = new HashMap<>();
+
+        for (Map.Entry<VariableReferenceExpression, RowExpression> entry : sourceAssignments.entrySet()) {
+            VariableReferenceExpression outputVariable = entry.getKey();
+            SpecialFormExpression ifExpr = (SpecialFormExpression) entry.getValue();
```

Comment:
- spell out ifExpression
- when we build `sourceAssignments`, we didn't filter on IF-only expressions. So there could be other types of expressions here. We probably wanna reuse some code in `ShouldRewriteAggregation`. Worth adding a test to cover this.

Reply:
sourceAssignments only has the references in the aggregations filtered by shouldRewriteAggregation. So it only has IF expressions. testFireForOneOfTwoAggregations covers this case.